password-cli/app/Main.hs

```diff
@@ -0,0 +1,177 @@
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards #-}
+
+module Main (main) where
+
+import Control.Monad (unless, void)
+import qualified Data.Password.Argon2 as Argon2
+import Data.Password.Bcrypt (PasswordCheck (..))
+import qualified Data.Password.Bcrypt as Bcrypt
+import qualified Data.Password.PBKDF2 as PBKDF2
+import qualified Data.Password.Scrypt as Scrypt
+import Data.Password.Types
+import Data.Text (Text)
+import qualified Data.Text.IO as T
+import Options.Applicative
+import System.Exit (exitFailure)
+import System.IO (IOMode(ReadMode), stdin, withFile)
+
+main :: IO ()
+main = execParser cliOpts >>= runCmd
+
+data Cmd
+  = HashCmd HashOpts
+  | CheckCmd CheckOpts
+  | HelpCmd (Maybe String)
+
+data HashOpts = HashOpts {
+    password :: Either FilePath Bool,
+    hashAlgorithm :: HashAlgorithm
+  }
+
+data CheckOpts = CheckOpts {
+    hash :: Either FilePath Text,
+    password :: Either FilePath Bool,
+    checkAlgorithm :: CheckAlgorithm
+  }
+
+data HashAlgorithm
+  = PBKDF2HashAlgo PBKDF2.PBKDF2Params
+  | BcryptHashAlgo Int
+  | ScryptHashAlgo Scrypt.ScryptParams
+  | Argon2HashAlgo Argon2.Argon2Params
+
+data CheckAlgorithm
+  = PBKDF2CheckAlgo
+  | BcryptCheckAlgo
+  | ScryptCheckAlgo
+  | Argon2CheckAlgo
+
+cliOpts :: ParserInfo Cmd
+cliOpts = info commandsParser (fullDesc <> header "password CLI usage -")
```

Comment:
The header only shows up if you do `password-cli --help`, and at our company we keep the `header` as the name of the program with the version affixed at the end:

`password-cli v0.1.0.0`

You can get the version from `import Paths_password_cli (version)` if you put that in `other-modules:` in the Cabal file.
It'll look something like this:
```bash
user@laptop-420 $ password-cli --help
password CLI v0.1.0.0

Usage: password-cli COMMAND
```

It'd also be nice that if the executable is run without ANYTHING behind it, it'd default to `--help`.

Reply:
> It'd also be nice that if the executable is run without ANYTHING behind it, it'd default to --help.

I'm not sure ho to do it

Reply:
In the `main` function, first use `System.Environment.getArgs` to check if there are no arguments.
And if there are none (i.e. the `[String]` is `[]`) you run `execParserPure` like in the `help` command, and only provide `--help` as argument.

At least, I think that should work as intended, right?